waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/PanopticOperationHandler.java

```diff
@@ -188,4 +192,38 @@
     // `Future.get` currently does not do that, but this is safe if implementation changes in the future
     return Math.max(1, requestTimeout + maxLatency);
   }
+
+  /**
+   * Implements {@link HMSHandler#get_all_functions()} over multiple metastores
+   *
+   * @return GetAllFunctionsResponse (db's from federated metastores will be prefixed if necessary)
+   */
+  public abstract GetAllFunctionsResponse getAllFunctions();
+
+  protected GetAllFunctionsResponse getAllFunctions(List<DatabaseMapping> databaseMappings) {
+    List<GetAllFunctionsRequest> allRequests = new ArrayList<>();
+
+    for (DatabaseMapping mapping : databaseMappings) {
+      GetAllFunctionsRequest getAllFunctionsRequest = new GetAllFunctionsRequest(mapping);
+      allRequests.add(getAllFunctionsRequest);
+    }
+
+    List<GetAllFunctionsResponse> responses = executeRequests(allRequests, GET_ALL_FUNCTIONS_TIMEOUT,
+        "Got exception fetching get_all_functions: {}");
+    if (responses.isEmpty()) {
+      return new GetAllFunctionsResponse();
+    }
+    GetAllFunctionsResponse result = new GetAllFunctionsResponse(responses.get(0));
+    responses.stream().skip(1).forEach(response -> addFunctionsToResult(response, result));
```

Comment:
The `1` confused me, but then I realised the first response is being added through the constructor in the line before. Could also use the default constructor and flatmap it, I _think_ this works:

```responses.stream().flatMap(response -> response.getFunctions().stream()).forEach(function -> result.addToFunctions(function));```

Then wouldn't need to method below. I think null functions can also be filtered too.

Reply:
Wouldn't that add the first response twice though? Because `result` already has the first response.

Reply:
Yes sorry, this is only with the default constructor. Like so:

```java
GetAllFunctionsResponse result = new GetAllFunctionsResponse();
responses.stream()
    .flatMap(response -> response.getFunctions()
        .stream().filter(functions -> functions != null))
    .forEach(function -> result.addToFunctions(function));
```

Reply:
I see, thanks!
Although, if the default constructor is used, I think it's easier to understand what `addFunctionsToResult` does without looking at the implementation instead of looking at 
`flatMap(response -> response.getFunctions().stream().filter(functions -> functions != null))`

I'm not saying the `flatMap` is hard to understand, just that the function describes what it does pretty well.

Reply:
Even without the `flatmap` I don't think the `.skip(1)` is necessary, can still just use the default constructor with no `skip`. But yes, true, the method name makes it easier to convey the meaning. Happy with either way :)

Reply:
I want to use the copy constructor just in case anything else besides the Functions are copied there. I want it to be like: Create a new response like the response from the primary (first one) then add all functions to this new response.

I'll have a closer look at Max suggestion not sure if I find it easier to read.

Reply:
@max-jacobs  your proposal filter out nulls from the functions list but the actual list can be null.

This I think would be same as my original code. Using `response.isSetFunctions()` but that's the same as `response.getFunctions != null`
```
    responses
        .stream()
        .skip(1)
        .filter(response -> response.isSetFunctions())
        .flatMap(response -> response.getFunctions().stream())
        .forEach(function -> result.addToFunctions(function));
```

Reply:
That looks good. If you don't think its easier to read then by all means stick with whats there currently. Happy with either way, I just love a good stream :)

Reply:
I'm going for it!